File: xkcd/src/main/java/xyz/jienan/xkcd/extra/ExtraPagerAdapter.java
package xyz.jienan.xkcd.extra;

import android.text.TextUtils;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

import java.util.List;

import xyz.jienan.xkcd.extra.fragment.SingleExtraFragment;
import xyz.jienan.xkcd.extra.fragment.SingleExtraWebViewFragment;
import xyz.jienan.xkcd.home.base.BaseStatePagerAdapter;
import xyz.jienan.xkcd.model.ExtraComics;

public class ExtraPagerAdapter extends BaseStatePagerAdapter {

    private List<ExtraComics> extraComicsList;

    public ExtraPagerAdapter(FragmentManager fm) {
        super(fm);
    }

    @Override
    public Fragment getItem(int position) {
        ExtraComics extraComics = extraComicsList.get(position);
        if (extraComics.links != null && !extraComics.links.isEmpty() && !TextUtils.isEmpty(extraComics.links.get(0))) {
            SingleExtraWebViewFragment fragment = SingleExtraWebViewFragment.Companion.newInstance(extraComics);
            fragmentsMap.put(position + 1, fragment);
            return fragment;
        }
        SingleExtraFragment fragment = SingleExtraFragment.Companion.newInstance(position + 1);
        fragmentsMap.put(position + 1, fragment);
        return fragment;
    }

    public void setEntities(List<ExtraComics> extraComics) {
        this.extraComicsList = extraComics;
    }
}

File: xkcd/src/main/java/xyz/jienan/xkcd/extra/ExtraPagerAdapter.kt
package xyz.jienan.xkcd.extra

import androidx.fragment.app.Fragment
import xyz.jienan.xkcd.extra.fragment.SingleExtraFragment
import xyz.jienan.xkcd.extra.fragment.SingleExtraWebViewFragment
import xyz.jienan.xkcd.home.base.BaseStatePagerAdapter
import xyz.jienan.xkcd.model.ExtraComics

class ExtraPagerAdapter(fragment: Fragment) : BaseStatePagerAdapter(fragment) {

    private var extraComicsList: List<ExtraComics>? = null

    override fun createFragment(position: Int): Fragment {
        val realPosition = position.coerceIn(0, extraComicsList!!.size - 1)
        val extraComics = extraComicsList!![realPosition]

        return if (extraComics.isWebExtra()) {
            SingleExtraWebViewFragment.newInstance(extraComics)
        } else {
            SingleExtraFragment.newInstance(realPosition + 1)
        }
    }

    fun setEntities(extraComics: List<ExtraComics>?) {
        extraComicsList = extraComics
    }

    override fun getItemCount(): Int = extraComicsList?.size ?: 0

    private fun ExtraComics.isWebExtra(): Boolean {
        return !links?.get(0).isNullOrBlank()
    }
}